File: app/src/main/java/yoyo/app/android/com/yoyoapp/Trip/profile/profile/ProfileFragment.java
package yoyo.app.android.com.yoyoapp.Trip.profile.profile;


import android.Manifest;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.Bundle;
import android.util.Base64;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.ViewModelProviders;
import com.karumi.dexter.Dexter;
import com.karumi.dexter.MultiplePermissionsReport;
import com.karumi.dexter.PermissionToken;
import com.karumi.dexter.listener.PermissionRequest;
import com.karumi.dexter.listener.multi.MultiplePermissionsListener;
import com.theartofdev.edmodo.cropper.CropImage;
import com.theartofdev.edmodo.cropper.CropImageView;
import de.hdodenhof.circleimageview.CircleImageView;
import org.json.JSONException;
import org.json.JSONObject;
import yoyo.app.android.com.yoyoapp.DataModels.User;
import yoyo.app.android.com.yoyoapp.Flight.Dialog.LanguageDialogFragment;
import yoyo.app.android.com.yoyoapp.Flight.Profile.*;
import yoyo.app.android.com.yoyoapp.MainActivity;
import yoyo.app.android.com.yoyoapp.R;
import yoyo.app.android.com.yoyoapp.Trip.Utils.UserSharedManager;
import yoyo.app.android.com.yoyoapp.Trip.authentication.AuthenticationActivity;
import yoyo.app.android.com.yoyoapp.Trip.profile.SignOutFragment;
import yoyo.app.android.com.yoyoapp.Trip.profile.editProfile.EditProfileFragment;
import yoyo.app.android.com.yoyoapp.Trip.profile.travellerCompanion.TravellerCompanionFragment;
import java.io.ByteArrayOutputStream;
import java.io.FileNotFoundException;
import java.io.InputStream;
import java.util.List;
import static android.app.Activity.RESULT_OK;


public class ProfileFragment extends Fragment {

    private FragmentManager fragmentManager;
    private User user;
    private TextView nameTextview, languageTextview;
    private ImageView circleImageView;
    private UserSharedManager userSharedManager;
    private boolean isSignedIn = true;
    private ProfileViewModel profileViewModel;
    private ImageView  travellerCompanionImageview, rulesImageview, editProfileImageview ,signoutImageview, languageImageView, aboutImageview;
    private View view;
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_profile_trip, container, false);

        init();
        checkingSignIn();
        retrieveProfileData();

        travellerCompanionImageview.setOnClickListener(v -> {if(isSignedIn) setupTravellerCompanionPage();else popUpSignInSignUpActivity();});
        rulesImageview.setOnClickListener(v -> setupRulesPage());
        editProfileImageview.setOnClickListener(v -> {if(isSignedIn) setupEditProfilePage(); else popUpSignInSignUpActivity();});
        signoutImageview.setOnClickListener(v -> { if (isSignedIn) setupSignOutPage(); else popUpSignInSignUpActivity();});
        aboutImageview.setOnClickListener(v -> setupAboutPage());
        languageImageView.setOnClickListener(v -> setupLanguage());


        return view;
    }


    private void setupAboutPage() {
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.replace(((MainActivity)getActivity()).getCurrentContainer(),new AboutFragment()).addToBackStack(String.valueOf(((MainActivity)getActivity()).getCurrentContainer()));
        fragmentTransaction.commit();
    }

    // show user the language dialog fragment
    private void setupLanguage() {
        LanguageDialogFragment languageDialogFragment = new LanguageDialogFragment();
        languageDialogFragment.show(getFragmentManager(), "setting language");
    }

    private void setupSignOutPage() {
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.replace(((MainActivity)getActivity()).getCurrentContainer(),new SignOutFragment()).addToBackStack(String.valueOf(((MainActivity)getActivity()).getCurrentContainer()));
        fragmentTransaction.commit();
    }

    private void setupEditProfilePage() {
            FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
            fragmentTransaction.replace(((MainActivity)getActivity()).getCurrentContainer(),new EditProfileFragment()).addToBackStack(String.valueOf(((MainActivity)getActivity()).getCurrentContainer()));
            fragmentTransaction.commit();
    }

    private void setupRulesPage() {
        FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
        fragmentTransaction.replace(((MainActivity)getActivity()).getCurrentContainer(),new RuleFragment()).addToBackStack(String.valueOf(((MainActivity)getActivity()).getCurrentContainer()));
        fragmentTransaction.commit();
    }

    private void setupTravellerCompanionPage() {
            FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
            fragmentTransaction.replace(((MainActivity)getActivity()).getCurrentContainer(),new TravellerCompanionFragment()).addToBackStack(String.valueOf(((MainActivity)getActivity()).getCurrentContainer()));
            fragmentTransaction.commit();
    }

    private void retrieveProfileData() {
        User user = userSharedManager.getUser();
        if (user.getFirstName().equals(""))
        {
            //todo set profile data
//            apiServiceFlight.retrieveProfileData(new ApiServiceFlight.OnProfileDataRecieved() {
//                @Override
//                public void onRecieved(User user) {
//                    if (user!=null)
//                    nameTextview.setText(user.getFirstName() + " " + user.getLastName());
//                }
//            });
        }
    }

    // if the user has not signed in yet, pop up the signup activity
    private void checkingSignIn() {
        user = userSharedManager.getUser();
        if (userSharedManager.getToken().equals(""))
        {
            popUpSignInSignUpActivity();
            isSignedIn = false;
        }
        else
        {
            setupViewsFromSharedPref();
        }
    }

    private void setupViewsFromSharedPref() {
        nameTextview.setText(user.getFirstName()+" "+ user.getLastName());
        if( !user.getProfilePicture().equalsIgnoreCase("") ){
            byte[] b = Base64.decode(user.getProfilePicture(), Base64.DEFAULT);
            Bitmap bitmap = BitmapFactory.decodeByteArray(b, 0, b.length);
            circleImageView.setImageBitmap(bitmap);
        }
        isSignedIn = true;
    }

    // setup signup activity
    private void popUpSignInSignUpActivity() {
        startActivity(new Intent(getActivity(), AuthenticationActivity.class));
        getActivity().overridePendingTransition(R.anim.slide_up,  R.anim.no_animation);
    }


    private void init() {
        profileViewModel = ViewModelProviders.of(getActivity()).get(ProfileViewModel.class);
        rulesImageview = view.findViewById(R.id.iv_profile_rules);
        travellerCompanionImageview = view.findViewById(R.id.iv_profile_traveller_companion);
        editProfileImageview = view.findViewById(R.id.iv_profile_edit_profile);
        nameTextview = view.findViewById(R.id.tv_profile_name);
        signoutImageview = view.findViewById(R.id.iv_profile_signout);
        fragmentManager = getFragmentManager();
        userSharedManager = new UserSharedManager(getContext());
        languageImageView = view.findViewById(R.id.iv_profile_language);
        aboutImageview = view.findViewById(R.id.iv_profile_aboout);
        languageTextview = view.findViewById(R.id.tv_profile_language);
        circleImageView = view.findViewById(R.id.iv_profile);
    }




}

File: app/src/main/java/yoyo/app/android/com/yoyoapp/trip/profile/profile/ProfileFragment.kt
package yoyo.app.android.com.yoyoapp.trip.profile.profile

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.lifecycle.Observer
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.fragment_profile.view.*
import yoyo.app.android.com.yoyoapp.DataModels.User
import yoyo.app.android.com.yoyoapp.Flight.Dialog.LanguageDialogFragment
import yoyo.app.android.com.yoyoapp.Flight.Profile.AboutFragment
import yoyo.app.android.com.yoyoapp.Flight.Profile.RuleFragment
import yoyo.app.android.com.yoyoapp.MainActivity
import yoyo.app.android.com.yoyoapp.MainActivityViewModel
import yoyo.app.android.com.yoyoapp.R
import yoyo.app.android.com.yoyoapp.trip.Utils.UserSharedManager
import yoyo.app.android.com.yoyoapp.trip.profile.SignOutFragment
import yoyo.app.android.com.yoyoapp.trip.profile.editProfile.EditProfileFragment
import yoyo.app.android.com.yoyoapp.trip.profile.travellerCompanion.TravellerCompanionFragment

class ProfileFragment : Fragment() {
    private val mainActivityViewModel by activityViewModels<MainActivityViewModel>()

    private lateinit var userSharedManager: UserSharedManager
    private var user: User? = null

    private var isSignedIn = false

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val res = inflater.inflate(R.layout.fragment_profile_trip, container, false)

        init()
        observe(res)
        checkSignIn()
        loadViewsIfSignedIn(res)
        setupClickListeners(res)

        return res
    }

    private fun observe(res: View) {
        observeUser(res)
    }

    private fun observeUser(res: View) {
        mainActivityViewModel.isUserSignedIn.observe(this, Observer {
            isSignedIn = it

            if (it) {
                user = userSharedManager.user
                res.tv_profile_name.text = "${user?.firstName} ${user?.lastName}"
            }
        })
    }

    private fun init() {
        userSharedManager = UserSharedManager(context)
        user = userSharedManager.user
    }

    private fun checkSignIn() {
        isSignedIn = userSharedManager.token != ""
    }

    private fun loadViewsIfSignedIn(res: View) {
        if (isSignedIn) {
            res.tv_profile_name.text = "${user?.firstName} ${user?.lastName}"
            if (!user?.profilePicture.equals("", ignoreCase = true)) {
                Picasso.with(context)
                    .load(user?.profilePicture)
                    .into(res.iv_profile)
            }
        }
    }

    private fun setupClickListeners(res: View) {
        setupTravellerCompanionClickListener(res)
        setupRulesClickListener(res)
        setupEditProfileClickListener(res)
        setupSignOutClickListener(res)
        setupAboutClickListener(res)
        setupLanguageClickListener(res)
    }

    private fun setupTravellerCompanionClickListener(res: View) {
        res.iv_profile_traveller_companion.setOnClickListener {
            if (isSignedIn)
                showTravellerCompanionPage()
            else
                popUpSignInSignUpActivity()
        }
    }

    private fun showTravellerCompanionPage() {
        (activity as MainActivity).showFragment(this, TravellerCompanionFragment(), "", false)
    }

    private fun popUpSignInSignUpActivity() {
        (activity as MainActivity).popUpSignInSignUpActivity()
    }

    private fun setupRulesClickListener(res: View) {
        res.iv_profile_rules.setOnClickListener { showRulesPage() }
    }

    private fun showRulesPage() {
        (activity as MainActivity).showFragment(this, RuleFragment(), "", false)
    }

    private fun setupEditProfileClickListener(res: View) {
        res.iv_profile_edit_profile.setOnClickListener {
            if (isSignedIn)
                showEditProfilePage()
            else
                popUpSignInSignUpActivity()
        }
    }

    private fun showEditProfilePage() {
        (activity as MainActivity).showFragment(this, EditProfileFragment(), "", false)
    }

    private fun setupSignOutClickListener(res: View) {
        res.iv_profile_signout.setOnClickListener {
            if (isSignedIn)
                showSignOutPage()
            else
                popUpSignInSignUpActivity()
        }
    }

    private fun showSignOutPage() {
        (activity as MainActivity).showFragment(this, SignOutFragment(), "", false)
    }

    private fun setupAboutClickListener(res: View) {
        res.iv_profile_aboout.setOnClickListener { showAboutPage() }
    }

    private fun showAboutPage() {
        (activity as MainActivity).showFragment(this, AboutFragment(), "", false)
    }

    private fun setupLanguageClickListener(res: View) {
        res.iv_profile_language.setOnClickListener {
            showLanguagePage()
        }
    }

    private fun showLanguagePage() {
        fragmentManager?.let {
            LanguageDialogFragment().show(it, "setting language")
        }
    }
}
